File: task1.py
from pyspark import SparkContext
import json
import sys
import time
import re


class Task1:
    def __init__(self):
        self.sc = SparkContext.getOrCreate()
        self.sc.setLogLevel("ERROR")
        self.input_file = sys.argv[1]
        self.output_file = sys.argv[2]
        self.stopwords_file = sys.argv[3]
        self.year = sys.argv[4]
        self.top_m = int(sys.argv[5])
        self.top_n = int(sys.argv[6])
        print("====== LOGS =====")
        print("Input file: {}".format(self.input_file))
        print("Output file: {}".format(self.output_file))
        print("Stopwords file: {}".format(self.stopwords_file))
        print("Year: {}".format(self.year))
        print("top_m: {}".format(self.top_m))
        print("top_n: {}".format(self.top_n))
        self.output = {}

    @staticmethod
    def wrangled(review):
        punctuations = set(['(', '[', ',', '.', '!', '?', ':', ';', ']', ')'])
        ret = []
        for c in review:
            if (c.isalpha() or c == " ") and c not in punctuations:
                ret.append(c)
        return "".join(ret)

    def read_data(self):
        with open(self.stopwords_file) as f:
            self.stopwords = set(map(lambda x: x.strip(), f.readlines()))

        self.data = self.sc.textFile(self.input_file)\
                        .map(lambda line: json.loads(line))
        self.data.cache()
        self.reviewers = self.data.map(lambda review: (review["user_id"], 1))
        self.reviewers.cache()

        self.reviews = self.data.map(lambda review: review["text"])\
                                    .map(lambda review: Task1.wrangled(review))\
                                    .map(lambda line: line.lower())\
                        
    def count_reviews(self):
        self.output["A"] = self.data.count()
    
    def count_reviews_in_year(self, year):
        self.output["B"] = self.data.filter(lambda review: review["date"][:4] == year).count()
    
    def count_distinct_reviewers(self):
        self.output["C"] = self.reviewers.distinct().count()

    def top_reviewers(self, top_m):
        self.output["D"] = self.reviewers.reduceByKey(lambda x, y: x + y).sortBy(lambda x: (-x[1], x[0])).take(top_m)

    def frequent_words(self, stopwords, top_n):
        self.output["E"] = self.reviews\
            .flatMap(lambda review: review.split())\
            .filter(lambda word: word not in stopwords)\
                .map(lambda word: (word, 1))\
                    .reduceByKey(lambda x, y: x + y)\
                        .sortBy(lambda x: -x[1])\
                            .take(top_n)

    def write_output(self):
        print(self.output)
        with open(self.output_file, "w") as f:
            json.dump(self.output, f)


if __name__ == "__main__":
    start_time = time.time()
    runner = Task1()    
    runner.read_data()

    # Task A
    runner.count_reviews()

    # Task B
    runner.count_reviews_in_year(runner.year)

    # Task C
    runner.count_distinct_reviewers()

    # Task D
    runner.top_reviewers(runner.top_m)

    # Task E
    runner.frequent_words(runner.stopwords, runner.top_n)

    runner.write_output()
    end_time = time.time()
    
    print("Total time taken: {}".format(end_time - start_time))

File: hw1/src/main/scala/task1.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD
import org.json4s._
import org.json4s.jackson.JsonMethods._

import java.io.{PrintWriter, File}
import scala.io.Source


object task1 {
  def jsonify(review: JValue):(String, Int) = {
    implicit val formats = DefaultFormats
    return ((review \ "user_id").extract[String], 1)
  }

  def BHelper(review: JValue, year: String):Boolean = {
    implicit val formats = DefaultFormats
    val str: String = (review \ "date").extract[String]
    val pyear = str.substring(0, 4)
    return pyear == year
  }

  def wrangle(review: JValue): String = {
    val punctuations = Set('(', '[', ',', '.', '!', '?', ':', ';', ']', ')')

    var ret = new StringBuilder()
    implicit val formats = DefaultFormats

    for (chr: Char <- review.extract[String]) {
      if (punctuations.contains(chr)) {
        ret.append(' ')
      }
      else if (chr.isLetter || chr == ' ') {
        ret.append(chr)
      }
      else {
        ret.append(' ')
      }
    }

    return ret.toString().toLowerCase()
  }

  def A(data: org.apache.spark.rdd.RDD[JValue]): Long = {
    return data.count()
  }

  def B(data: org.apache.spark.rdd.RDD[JValue], year: String): Long = {
    return data.filter(x => BHelper(x, year)).count()
  }

  def C(data: org.apache.spark.rdd.RDD[(String, Int)]): Long = {
    return data.distinct().count()
  }

  def D(data: org.apache.spark.rdd.RDD[(String, Int)], top_m: Int): Array[List[Any]] = {
    return data.reduceByKey(_+_).sortBy(x => (-x._2, x._1)).take(top_m).map(item => item.productIterator.toList)
  }

  def E(data: RDD[String], stopwords: Set[String], top_n: Int): Array[String] = {
    return data.flatMap(review => review.split(" +"))
      .filter(word => !stopwords.contains(word))
      .map(word => (word, 1))
      .reduceByKey(_+_)
      .sortBy(x => -x._2)
      .take(top_n).map(_._1)
  }

  def main(args:Array[String]): Unit = {
    val sc = SparkContext.getOrCreate()
    sc.setLogLevel("ERROR")

    for (arg <- args) {
      println(arg)
    }

    val input_file = args(0)
    val output_file = args(1)
    val stopwords_file = args(2)
    val year = args(3)
    val top_m = args(4).toInt
    val top_n = args(5).toInt

    var output: Map[String, Any] = Map()

    val textRDD = sc.textFile(input_file)
    val data = textRDD.map(review => parse(review))
    data.cache()

    val reviewers = data.map(review => jsonify(review))
    reviewers.cache()

    val reviews = data.map(review => review \ "text")
      .map(line => wrangle(line))

    val stopwords = Source.fromFile(stopwords_file).getLines.toSet

    val a = A(data)
    val b = B(data, year)
    val c = C(reviewers)
    val d = D(reviewers, top_m)
    val e = E(reviews, stopwords, top_n)

    output += {"A"->a}
    output += {"B"->b}
    output += {"C"->c}
    output += {"D"->d}
    output += {"E"->e}

    implicit val formats = org.json4s.DefaultFormats
    val formatted_output = org.json4s.jackson.Serialization.write(output)

    println(formatted_output)

    val out = new PrintWriter(new File(output_file))
    out.write(formatted_output)
    out.close()
  }
}